File: lib/domain/pest/ports/pest_update_output_port.rb
# frozen_string_literal: true

module Domain
  module Pest
    module Ports
      class PestUpdateOutputPort
        def on_success(pest_entity)
          raise NotImplementedError, "Subclasses must implement on_success"
        end

        # @param failure_dto [Domain::Pest::Dtos::PestMasterFormFailure, Domain::Shared::Dtos::ReferenceFlagChangeDeniedFailure, Domain::Shared::Dtos::Error, Domain::Shared::Policies::PolicyPermissionDenied]
        def on_failure(failure_dto)
          raise NotImplementedError, "Subclasses must implement on_failure"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/pest/ports/pest_update_output_port.rs
use crate::pest::entities::PestEntity;
use crate::shared::dtos::{Error, ReferenceFlagChangeDeniedFailure};
use crate::shared::policies::policy_permission_denied::PolicyPermissionDenied;

pub trait PestUpdateOutputPort {
    fn on_success(&mut self, entity: PestEntity);
    fn on_failure(&mut self, error: UpdateFailure);
}

#[derive(Debug, Clone)]
pub enum UpdateFailure {
    Policy(PolicyPermissionDenied),
    Error(Error),
    ReferenceFlagChange(ReferenceFlagChangeDeniedFailure),
}
